lib/checkcondition.cpp

```diff
@@ -1188,12 +1187,13 @@ void CheckCondition::alwaysTrueFalse()
                 continue;
             if (Token::Match(tok, "! %num%|%bool%"))
                 continue;
+            if (Token::Match(tok, "&&|%oror%") && !Token::Match(tok->astOperand1(), "%char%") && !Token::Match(tok->astOperand2(), "%char%"))
+                continue;
 
             const bool constIfWhileExpression =
-                tok->astParent()
-                && Token::Match(tok->astParent()->astOperand1(), "if|while")
-                && !tok->isBoolean();
-            const bool constValExpr = Token::Match(tok, "%num%|%char%") && Token::Match(tok->astParent(),"%oror%|&&|?"); // just one number or char in boolean expression
+                tok->astTop() && Token::Match(tok->astTop()->astOperand1(), "if|while") &&
+                (Token::Match(tok->astParent(),"&&|%oror%") || Token::Match(tok->astParent()->astOperand1(), "if|while"));
```

Comment:
Please review the `constIfWhileExpression` expression. as I see it the last `Token::Match` seems redundant.